### DISCUSSION TITLE
Gatsby Source Wordpress Missing Data

### DISCUSSION THREAD

**Discussion by adkanojia** [gatsbyjs/gatsby#35491] [Help] [answered]
Hello community,

I need some help understanding the options of gatsby-source-wordpress plugin.
During development/build process I run into an error 

```
gatsby-source-wordpress  Request failed with status code 500

Your WordPress server is either overloaded or encountered a PHP error.

Error occurred while fetching nodes of the "Restaurant" type.

Enable WordPress debugging by adding the following to your wp-config.php file:

define("WP_DEBUG", true);
define("WP_DEBUG_LOG", true);
define("GRAPHQL_DEBUG", true);

(See https://wordpress.org/support/article/debugging-in-wordpress/ for more info)

Then run another build before checking your WordPress instance's debug.log file for errors.

If you don't see any errors in debug.log:

Try reducing the requestConcurrency for content updates or the previewRequestConcurrency for previews, and/or reducing the schema.perPage option:

{
  resolve: 'gatsby-source-wordpress',
  options: {
    schema: {
      perPage: 20, // currently set to 100
      requestConcurrency: 5, // currently set to 15
      previewRequestConcurrency: 2, // currently set to 5
    }
  },
}
```

I have the debug log enabled and the only error I see is about memory issue.
However, changing `requestConcurrency` or `previewRequestConcurrency` doesn't change anything. I still get the error.
Changing `perPage` to 20 fixes the error. Though with this change I don't get all data in my build.

Does setting `perPage` in schema also limits the data on your build?
I can't find the solution to this, only solution I'm left with is increasing the memory limit on my WP server.
I've tested this on a fresh gatsby install as well with same results.
Can't find any other documentation on `schema.perPage` option.

Any help will be appreciated.
Thanks

**Comment by adkanojia** [comment, marked as answer]
The issue was caused by a couple plugins which I was using for sorting/ordering

- [Post Types Order](https://wordpress.org/plugins/post-types-order/)
- [Category Order and Taxonomy Terms Order](https://wordpress.org/plugins/taxonomy-terms-order/)

These plugins has option to override db queries to fetch posts/categories in a specific order set by `menuOrder` and `termOrder`.
When you disable the option to override global queries, the missing data issue is resolved.

If you still want to get the posts in that order in your graphql query results, you can use
`(where: {orderby: {field: MENU_ORDER, order: ASC}})` for posts
and
`(where: {orderby: TERM_ORDER})` for terms

Hope this'll help anyone else struggling with same issue.

**Reply by jploskonka** [reply]
@adkanojia Thank you, thank you so much for adding this answer! I've been struggling for over a week to find a reason for missing posts in my gatsby builds. 

I'm using mentioned Post Types Order plugin. Now, depending on different `perPage` values, I was missing different posts. For example, with `perPage` set to `25` I was missing ~70 posts out of 542. When `perPage` was set to `100` I was missing about `40` posts etc.

What solved the issue for me was unticking this box in `Post Types order` settings in WordPress admin. You can get there by choosing `Settings -> Post Types Order` from left menu.
![Screenshot 2022-07-21 at 18 54 15](https://user-images.githubusercontent.com/669483/180270492-c5af342f-6485-40e0-9fa3-f793a37a223a.png)

Hope this helps someone else struggling with the issue :-)

**Reply by jploskonka** [reply]
Actually, after unticking the `auto sort` box I started getting all the posts but was not able to get them in the correct order. In `gatsby-source-wordpress` graphql schema I couldn't find any field that would reflect `MENU_ORDER`. 

I gave up on this and instead went with adding the snippet mentioned in this link: https://www.wpgraphql.com/recipes/fix-pagination-conflict-with-the-post-type-order-plugin/ to my `functions.php` file in WordPress and ticking `Auto Sort` box in `Post Types` plugin settings. Now, pagination when fetching posts works just fine and I'm getting all results sorted properly.

Here's also a snippet from the mentioned link:

``` PHP
add_action( 'pre_get_posts', function () {

	// access the custom post type order class
	global $CPTO;

	// if WPGraphQL isn't active or CPTUI isn't active, carry on and bail early
	if ( ! function_exists( 'is_graphql_request' ) || ! is_graphql_request() || ! $CPTO ) {
		return;
	}

	// Remove the Post Type Order plugin's "posts_orderby" filter for WPGraphQL requests
	// This filter hooks in too late and modifies the SQL directly so WPGraphQL
	// can't properly map the orderby args to generate the SQL for proper pagination
	remove_filter( 'posts_orderby', [ $CPTO, 'posts_orderby' ], 99 );

	// Add a filter
	add_filter( 'graphql_post_object_connection_query_args', function ( $args, $source, $input_args, $context, $info ) {

		$orderby = [];

		// If the connection has explicit orderby arguments set,
		// use them
		if ( ! empty( $input_args['where']['orderby'] ) ) {
			return $args;
		}

		// Else use any orderby args set on the WP_Query
		if ( isset( $args['orderby'] ) ) {
			$orderby = [];

			if ( is_string( $args['orderby'] ) ) {
				$orderby[] = $args['orderby'];
			} else {
				$orderby = $args['orderby'];
			}
		}

		$orderby['menu_order'] = ! empty( $input_args['last'] ) ? 'DESC' : 'ASC';
		$args['orderby']       = $orderby;

		return $args;
	}, 10, 5 );

});
```

**Reply by adkanojia** [reply]
Hi @jploskonka 
Glad you fixed the issue.

I remember, I used to get the `MENU ORDER` field in graphql as well, so I sorted the posts in graphql query instead of adding the snippet in `functions.php`
Just check that in `graphql playground` that might help.

Cheers
